internal/lib/stats/stats_server.go

```diff
@@ -0,0 +1,339 @@
+		if err != nil {
+			logrus.Errorf("Error getting container stats %s: %v", c.ID(), err)
+			continue
+		}
+		ss.populateWritableLayer(cStats, c)
+		if oldcStats, ok := ss.ctrStats[c.ID()]; ok {
+			updateUsageNanoCores(oldcStats.CPU, cStats.CPU)
+		}
+	}
+	sandboxStats.Containers = containerStats
+	if old, ok := ss.sboxStats[sb.ID()]; ok {
+		updateUsageNanoCores(old.CPU, sandboxStats.CPU)
+	}
+	ss.sboxStats[sb.ID()] = sandboxStats
+	return sandboxStats
+}
+
+// updateContainer calls into the runtime handler to update the container stats,
+// as well as populates the writable layer by calling into the container storage.
+// If this container already existed in the stats server, the CPU nano cores are calculated as well.
+func (ss *StatsServer) updateContainer(c *oci.Container, sb *sandbox.Sandbox) *types.ContainerStats {
+	if c == nil || sb == nil {
+		return nil
+	}
+	if c.StateNoLock().Status == oci.ContainerStateStopped {
+		return nil
+	}
+	cStats, err := ss.Runtime().ContainerStats(context.TODO(), c, sb.CgroupParent())
+	if err != nil {
+		logrus.Errorf("Error getting container stats %s: %v", c.ID(), err)
+		return nil
+	}
+	ss.populateWritableLayer(cStats, c)
+	if oldcStats, ok := ss.ctrStats[c.ID()]; ok {
+		updateUsageNanoCores(oldcStats.CPU, cStats.CPU)
+	}
+	ss.ctrStats[c.ID()] = cStats
+	return cStats
+}
+
+// updateUsageNanoCores calculates the usage nano cores by averaging the CPU usage between the timestamps
+// of the old usage and the recently gathered usage.
+func updateUsageNanoCores(old, current *types.CPUUsage) {
+	if old == nil || current == nil || old.UsageCoreNanoSeconds == nil || current.UsageCoreNanoSeconds == nil {
+		return
+	}
+
+	nanoSeconds := current.Timestamp - old.Timestamp
+
+	usageNanoCores := uint64(float64(current.UsageCoreNanoSeconds.Value-old.UsageCoreNanoSeconds.Value) /
+		float64(nanoSeconds) * float64(time.Second/time.Nanosecond))
+
+	current.UsageNanoCores = &types.UInt64Value{
+		Value: usageNanoCores,
+	}
+}
+
+// populateWritableLayer attempts to populate the container stats writable layer.
+func (ss *StatsServer) populateWritableLayer(stats *types.ContainerStats, container *oci.Container) {
+	writableLayer, err := ss.writableLayerForContainer(container)
+	if err != nil {
+		logrus.Errorf("%v", err)
+	}
+	stats.WritableLayer = writableLayer
+}
+
+// writableLayerForContainer gathers information about the container's writable layer.
+// It does so by calling into the GraphDriver's endpoint to get the UsedBytes and InodesUsed.
+func (ss *StatsServer) writableLayerForContainer(container *oci.Container) (*types.FilesystemUsage, error) {
+	writableLayer := &types.FilesystemUsage{
+		Timestamp: time.Now().UnixNano(),
+		FsID:      &types.FilesystemIdentifier{Mountpoint: container.MountPoint()},
+	}
+	driver, err := ss.Store().GraphDriver()
+	if err != nil {
+		return writableLayer, errors.Wrapf(err, "Unable to get graph driver for disk usage for container %s", container.ID())
```

Comment:
Not for this PR, but in general: at some point we should ditch pkg/errors and switch to go native error wrapping (i.e. `fmt.Errorf(".... %w", err)`.

Reply:
agreed!